util/linuxfw/linuxfw.go

```diff
@@ -2,10 +2,254 @@
 // SPDX-License-Identifier: BSD-3-Clause
 
 // Package linuxfw returns the kind of firewall being used by the kernel.
+
+//go:build linux
+
 package linuxfw
 
-import "errors"
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/tailscale/netlink"
+	"tailscale.com/types/logger"
+)
+
+type ProtoVersion int
+
+const (
+	IPv4 ProtoVersion = iota
+	IPv6
+)
+
+type MatchDecision int
+
+const (
+	Accept MatchDecision = iota
+	Masq
+)
+
+// The following bits are added to packet marks for Tailscale use.
+//
+// We tried to pick bits sufficiently out of the way that it's
+// unlikely to collide with existing uses. We have 4 bytes of mark
+// bits to play with. We leave the lower byte alone on the assumption
+// that sysadmins would use those. Kubernetes uses a few bits in the
+// second byte, so we steer clear of that too.
+//
+// Empirically, most of the documentation on packet marks on the
+// internet gives the impression that the marks are 16 bits
+// wide. Based on this, we theorize that the upper two bytes are
+// relatively unused in the wild, and so we consume bits 16:23 (the
+// third byte).
+//
+// The constants are in the iptables/iproute2 string format for
+// matching and setting the bits, so they can be directly embedded in
+// commands.
+const (
+	// The mask for reading/writing the 'firewall mask' bits on a packet.
+	// See the comment on the const block on why we only use the third byte.
+	//
+	// We claim bits 16:23 entirely. For now we only use the lower four
+	// bits, leaving the higher 4 bits for future use.
+	TailscaleFwmarkMask    = "0xff0000"
+	TailscaleFwmarkMaskNeg = "0xff00ffff"
+	TailscaleFwmarkMaskNum = 0xff0000
+
+	// Packet is from Tailscale and to a subnet route destination, so
+	// is allowed to be routed through this machine.
+	TailscaleSubnetRouteMark    = "0x40000"
+	TailscaleSubnetRouteMarkNum = 0x40000
+	// This one is same value but padded to even number of digit, so
+	// hex decoding can work correctly.
+	TailscaleSubnetRouteMarkHexStr = "0x040000"
+
+	// Packet was originated by tailscaled itself, and must not be
+	// routed over the Tailscale network.
+	//
+	// Keep this in sync with tailscaleBypassMark in
+	// net/netns/netns_linux.go.
+	TailscaleBypassMark    = "0x80000"
+	TailscaleBypassMarkNum = 0x80000
+)
+
+// RouteTable is a Linux routing table: both its name and number.
+// See /etc/iproute2/rt_tables.
+type RouteTable struct {
+	Name string
+	Num  int
+}
+
+// IpCmdArg returns the string form of the table to pass to the "ip" command.
+func (rt RouteTable) IpCmdArg() string {
+	if rt.Num >= 253 {
+		return rt.Name
+	}
+	return strconv.Itoa(rt.Num)
+}
+
+var routeTableByNumber = map[int]RouteTable{}
```

Comment:
Are you certain that this is only ever accessed by a single goroutine? If not, it will need synchronization.

Do we need to use route table names, or can we exclusively use numbers? If we can exclusively use numbers, can we remove the need for this entirely?

Reply:
This I what I took from router_linux, but it appears to me that it's no longer needed in linuxfw package. Also it seems like it's more reasonable to be in router_linux just looking at it's names and references, so I will just return this to router_linux.go. 

I'm not sure if there will be more goroutine accessing but I didn't really change the logic so I guess @bradfitz  was certain about what this meant. I can do a further investigation

Reply:
At least in the old code in router_linux.go that's a global map that's only written to during `init`, not at runtime.